File: app/src/main/java/com/example/thenewipad/function/GetPtxBusJson.java
package com.example.thenewipad.function;

import android.icu.text.SimpleDateFormat;
import android.icu.util.Calendar;
import android.icu.util.TimeZone;
import android.os.Build;

import androidx.annotation.RequiresApi;

import com.example.thenewipad.formatFolder.BusArriveStop;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.lang.reflect.Type;
import java.net.HttpURLConnection;
import java.net.ProtocolException;
import java.net.URL;
import java.security.SignatureException;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.zip.GZIPInputStream;

public class GetPtxBusJson {
    String url;
    String response = "";
    public GetPtxBusJson(String url){
        this.url = url;
        this.response = response;
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public String request() {
        HttpURLConnection connection = null;
        String str;
        String APPID = "8af3b6907d5c443d90b6bd2f18361c15";
        String APPKey = "6f5mGu2y_kH-MaeZ2NKlS0_MDzI";
        String xdate = getServerTime();
        String SignDate = "x-date: " + xdate;
        String respond = "";
        String Signature = "";
        try {
            Signature = HMAC_SHA1.Signature(SignDate, APPKey);
        } catch (SignatureException e1) {
            // TODO Auto-generated catch block
            e1.printStackTrace();
        }
        System.out.println("Signature :" + Signature);
        String sAuth = "hmac username=\"" + APPID + "\", algorithm=\"hmac-sha1\", headers=\"x-date\", signature=\""
                + Signature + "\"";
        System.out.println(sAuth);
        try {
            URL url = new URL(this.url);
            if ("https".equalsIgnoreCase(url.getProtocol())) {
                SslUtils.ignoreSsl();
            }
            connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.setRequestProperty("Authorization", sAuth);
            connection.setRequestProperty("x-date", xdate);
            connection.setRequestProperty("Accept-Encoding", "gzip");
            connection.setDoInput(true);
            //connection.setDoOutput(true);
            respond = connection.getResponseCode() + " " + connection.getResponseMessage();

            BufferedReader in;

            // 判斷來源是否為gzip
            if ("gzip".equals(connection.getContentEncoding())) {
                InputStreamReader reader = new InputStreamReader(new GZIPInputStream(connection.getInputStream()));
                in = new BufferedReader(reader);
            } else {
                InputStreamReader reader = new InputStreamReader(connection.getInputStream());
                in = new BufferedReader(reader);
            }
            // 返回的數據已經過解壓
            StringBuffer buffer = new StringBuffer();
            // 讀取回傳資料
            String line = "";
            while ((line = in.readLine()) != null) {
                response += (line + "\n");
            }
        } catch (ProtocolException e) {
            e.printStackTrace();
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        return response;
    }

    // 取得當下UTC時間
    public static String getServerTime() {
        Calendar calendar = Calendar.getInstance();
        SimpleDateFormat dateFormat = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss z", Locale.US);
        dateFormat.setTimeZone(TimeZone.getTimeZone("GMT"));
        return dateFormat.format(calendar.getTime());
    }

}

File: app/src/main/java/com/example/thenewipad/function/GetPtxBusJson.kt
package com.example.thenewipad.function

import android.icu.text.SimpleDateFormat
import android.icu.util.Calendar
import android.icu.util.TimeZone
import android.os.Build
import androidx.annotation.RequiresApi
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.ProtocolException
import java.net.URL
import java.security.SignatureException
import java.util.*
import java.util.zip.GZIPInputStream


/* 這邊請不要管他!!!!
 * 這個不是我寫的
 * Kotlin分析比較快 */


open class GetPtxBusJson() {
    @kotlin.jvm.JvmField
    var response = ""

    fun request(url: String): String {
        var connection: HttpURLConnection? = null
        val APPID = "8af3b6907d5c443d90b6bd2f18361c15"
        val APPKey = "6f5mGu2y_kH-MaeZ2NKlS0_MDzI"
        val xdate = serverTime
        val SignDate = "x-date: $xdate"
        var respond = ""
        var Signature = ""
        try {
            Signature = HMAC_SHA1.Signature(SignDate, APPKey).toString()
        } catch (e1: SignatureException) {
            e1.printStackTrace()
        }
        val sAuth = ("hmac username=\"" + APPID + "\", algorithm=\"hmac-sha1\", headers=\"x-date\", signature=\""
                + Signature + "\"")
        try {
            val url = URL(url)
            if ("https".equals(url.protocol, ignoreCase = true)) {
                SslUtils.ignoreSsl()
            }
            connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.setRequestProperty("Authorization", sAuth)
            connection.setRequestProperty("x-date", xdate)
            connection.setRequestProperty("Accept-Encoding", "gzip")
            connection.doInput = true
            //connection.setDoOutput(true);
            respond = connection.responseCode.toString() + " " + connection.responseMessage
            val `in`: BufferedReader

            // 判斷來源是否為gzip
            `in` = if ("gzip" == connection.contentEncoding) {
                val reader = InputStreamReader(GZIPInputStream(connection.inputStream))
                BufferedReader(reader)
            } else {
                val reader = InputStreamReader(connection.inputStream)
                BufferedReader(reader)
            }
            // 返回的數據已經過解壓
            val buffer = StringBuffer()
            // 讀取回傳資料
            var line = ""
            while (`in`.readLine().also { line = it } != null) {
                response += """
                    $line
                    """.trimIndent()
            }
        } catch (e: ProtocolException) {
            e.printStackTrace()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return response
    }
    companion object {
        // 取得當下UTC時間
        val serverTime: String
            get() {
                val calendar = Calendar.getInstance()
                val dateFormat = SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss z", Locale.US)
                dateFormat.timeZone = TimeZone.getTimeZone("GMT")
                return dateFormat.format(calendar.time)
            }
    }
}